### ISSUE TITLE
Handle chain reorgs

### ISSUE THREAD

**Issue by 0xbok** [ArcPay/ArcPay#30]
Scenarios:
- A promise is with a user, operator honors that promise, but then the chain reorgs -> operator can be slashed.
- For v2, let's say a user off-ramps and then chain reorgs -> system goes into bad debt.

**Comment by BlakeMScurr** [comment]
OK this is interesting, I haven't thought about reorgs much, so lmk if I'm missing important concepts!

Let's unpack scenario 1 a little. There are a few ways the operator can be slashed: if they are too slow to dequeue forced transactions or make new blocks; if they are too slow to give proof requests; if they haven't included a promised transaction in the appropriate block.
There's a simple solution to the third one: the operator should never update the state incorrectly on *any* for of the chain. The state can only look like the operator has broken a promise if the operator actually pushes a state update that breaks that promise (assuming that the ZKP works, and L1 has valid state transitions).
I think the time-dependent slashing is a lot trickier though. I think the operator just has to be very vigilant about making sure that they have responded to forcing etc on every possible fork in time.

Can you explain the second scenario a little more?

**Comment by 0xbok** [comment]
yes, if the operator is monitoring and updating each major fork, then this problem is handled. However, the UX becomes an issue, like what's the user balance etc. (that's a separate topic so we can branch it off later).

Scenario 2: by off-ramp i mean crypto -> fiat. let's say we add a feature where usdc can be withdrawn to bank directly from L2 without withdrawing to L1 first. This is a permanent change and after this if the chain reorgs, we can never recover that off-ramped money.

**Comment by 0xbok** [comment]
Circle added USDC native support for Arbitrum. will be good to check out how they handle L1 reorg.

**Comment by BlakeMScurr** [comment]
Regarding the UX question, I think we should basically show balances based on promises, and put a little lock beside transactions/balances once they're confirmed onchain. So the operator can actually maintain a ground truth of balances (namely, the ones that *will* make it into the onchain state), but if there's a reorg they might have to remove a little lock symbol, because the transaction is no longer locked in onchain. I'll put that stuff in the design!

My mental model for the fiat offramp is that it's not exactly built into the system, we just have an address, and when we receive tokens we send the fiat. So I don't think the *system* itself will go into bad debt. Also, once we have a signed transction from the user, we can always make sure it gets included, even if it gets reorged out. We just need to make sure the forcing wait period is long enough that we can execute the transaction before they send it somewhere else.

**Comment by 0xbok** [comment]
you're right! we can execute the promise again after reorg.
